### ISSUE TITLE
Parser with infix & adjacency

### ISSUE THREAD

**Issue by jonsterling** [RedPRL/sml-redprl#72]
No need to add custom infix notation just yet; but we do need to set up the `parcom` machinery for parsing built-in infix operators properly, as well as `ap(M;N)` by adjacency (`M N`).

We had this in JonPRL here: https://github.com/jonsterling/JonPRL/blob/f7445158007c9761a04602b1beaf47fa2990d710/src/syntax/syntax.sml#L112

**Comment by jonsterling** [comment]
I have set up the parcom machinery, but it seems like postfix & adjacency don't interact well. The following should work: `M.welp foo`, but this fails to parse; only `(M.welp) foo` works.

**Comment by jonsterling** [comment]
I wonder if we should find some way to trash the hand-written parser and use something like mlyacc or something. I'm not sure of the implications of this, and I don't know if we can support the syntax/sort-guided notation parsing, but it might be worth looking into. The hand-written parser is hard to maintain and easy to screw up.

**Comment by ayberkt** [comment]
I believe ML-Yacc can be set up to work with the syntax representation of `sml-typed-abts` but even a nicer option would be to use [BNFC](http://bnfc.digitalgrammars.com/).

It has the really nice features of generating grammar documentation from the BNF given and not making you deal with Lex and Yacc. Also, the grammar becomes much more maintainable since you just deal with a BNF representation.

Currently, it does not have a back-end for SML though it does have one for Ocaml (with OCaml-Yacc) which would be easy to adapt to SML. The problem is, I don't know if _this_ will work with the syntax representation (without giving up some of the benefits) because of the [labelled BNF](http://bnfc.digitalgrammars.com/LBNF-report.pdf) formalism it uses. But if using it for RedPRL is interesting I can try and see if it's possible to do the analogous thing in OCaml in a nice way and if that's the case, write an SML back-end.

**Comment by jonsterling** [comment]
@ayberkt This sounds really interesting! Do you think you could tell me a bit about your concerns with labelled BNF? I don't know anything about it...

**Comment by jonsterling** [comment]
By the way, in case it simplifies anything, I'm not necessarily hoping to develop a generated grammar for arbitrary ABT signatures, just the ones in RedPRL, since almost all our concrete syntax is custom.

**Comment by jonsterling** [comment]
@ayberkt Ah, I just looked at BNFC and it seems really really cool, but I'm not willing to make a Haskell tool a build dependency of RedPRL.

**Comment by ayberkt** [comment]
@jonsterling okay that's fair 😄. Then I'll see if I can get ML-Yacc working. Is the [grammar here](https://github.com/RedPRL/sml-redprl/blob/master/doc/definition.pdf) up to date?

**Comment by jonsterling** [comment]
@ayberkt Wonderful! 😄  By the way, my reasoning for not wanting Haskell dependencies isn't sectarian, but it just makes it harder to build RedPRL and difficult to predict what will happen to the build over the years.

The grammar for signatures in the Definition is up to date; I do not have a concrete grammar written down for expressions (such as terms, tactic scripts, etc.). If you end up working on this, let me know when you are ready for such a grammar and I can make one for you. Also feel free to make it up as you go (whatever's easiest!).

**Comment by jonsterling** [comment]
@ayberkt I've been thinking a bit more about this, and I had a few remarks to make about the parser (in case I hadn't mentioned these already):

1. The sort-directed / contextual parser that we have right now will likely be very difficult to achieve. I think we should perhaps just abandon that approach and try to come up with a simple grammar that is easy to implement.

2. The parser for the tactic language may be quite difficult to implement as-is; feel free to leave this to the end, and I can try and help make this a bit more sane.

**Comment by ayberkt** [comment]
@jonsterling great to see progress on this! I've looked at what you recently did and there seems to be changes to the grammar. Will it be further changing? If not I can continue working on getting the grammar complete in `new-syntax`. Also should we update the grammar in the definition as we go?

Now that I'm back to normal life, I believe I'll have much more time to work on this :smile:.

**Comment by jonsterling** [comment]
@ayberkt Hey! 😄 

To give a brief summary of what is happening in that branch:

1. I have had to generalize the ABT library a bit to account for "cubical syntax", because we are now implementing Carlo & Bob's cubical type theory.

2. In the process of doing so, I also designed a new abstract machine that will be much cleaner than the current one; one of the benefits is it does not require us to have such a complicated syntactic structure, so we can say goodbye to the current polarization into "values" and "continuations", linked together by "cuts". That structure will still exist, but only latently, and not reified into the actual design of our operator signatures.

3. As a result, a *lot* of stuff in RedPRL will have to change; luckily, because of the `Syntax` view, these changes will not affect the refiner. But basically everything that lives below that view (so, the syntax itself, the parser, the `Signature` apparatus) needs to be changed.

Because of this, I've been re-developing RedPRL's syntax in a new folder called `new-syntax`, in a new branch. There's a number of things that are in RedPRL's current syntax that I'm going to leave out for now, and I'm mostly interested in getting a minimal language working, and not worrying about maintaining parity with RedPRL's existing feature set (which is tiny anyway).

If you want to help out, that'd be awesome! We will need to parse things like signatures, tactic scripts, etc. in addition to what I have already implemented.

> I've looked at what you recently did and there seems to be changes to the grammar. Will it be further changing?

The grammar may be changing a bit, in order to accommodate the lack of sophistication possible under the mlyacc-style parsers. I'd like to keep things very simple for now. If there are changes to the grammar that conflict with the Definition, it would be great to make updates there! However, if it is easier, it's also permissible to just delete the relevant parts of the Definition.

I see the Definition as a design document that helps us propel the code further, so in some sense, if the code diverges from the Definition, then it has served its purpose. I would like to have a full specification of everything at some point, but today I'd be satisfied with having a working proof assistant. 😄
